### PR TITLE
Add combinators 'OnBegin' and 'OnEnded'

### LINKED ISSUE #527: Provide a way to run cleanup at the end of stream generation
I don't know how to write "cleanup" code that runs at the end of stream generation. For example (pseudocode):

```
return TypeCheck<Generator<int>>(Iterate({1, 2})
    >> Map([](int value) { return value; })
    >> Then([]() { LOG(INFO) << "All done!; })
);
```

The final `Then` statement changes the type of the resulting eventual chain such that it's no longer able to create a generator from the result.

This came up when writing a gRPC server handler for a bidirectional streaming RPC, where the server handler wanted to run some extra logic after sending a final request to a client.

### PR DESCRIPTION
**Pull request by benh**
Adds combinators for doing computation (possibly asynchronously) when a stream/generator begins or ends.

Fixes #527

### CODE REVIEW

**Review by ArthurBandaryk — APPROVED** [review]

File: eventuals/on-begin.h [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+#include "eventuals/stream.h"
+#include "eventuals/then.h"
+
+////////////////////////////////////////////////////////////////////////
+
+namespace eventuals {
+
+////////////////////////////////////////////////////////////////////////
+
+// Returns an eventual which will invoke the specified function when a
```
Comment (CodingCanuck) [inline eventuals/on-begin.h:13]: Drive-by: thanks for including such great documentation for these!
